### PR TITLE
Radial Menu Bandages

### PR DESCRIPTION
**Pull request by Hughgent**
## About The Pull Request

PR has been opened.
Things to do:
- [x] Add it to preferences for players to use/not use.
- [x] Check to make sure it works.
- [x] Get some sprites for the radial menu icons.

It works for surgery steps now.

## Why It's Good For The Game

People will be able to apply bandages/whathaveyou by using this radial menu.

## Changelog
:cl: Hughgent
add: Radial menu for medical bandages.
add: Preference for players to turn Radial menu on and off. Doctors may prefer off.
/:cl:

### CODE REVIEW

**Comment by Hughgent** [comment]
getting pretty snazzy
image

**Comment by Hughgent** [comment]
I bet this could work for surgery steps. I think I'd have to move my `radial_medical()` proc somewhere else.

**Comment by Hughgent** [comment]
Issue at the moment is that `show_radial_menu_persistent()` doesn't seem to be working as is.

Thus my intended function of the radial menu for bandages, where it menu stays open just allowing someone to click along the parts, is a pipe dream at the moment. Probably just not implementing it properly.

File: code/game/objects/items/stacks/medical.dm [hunk 1]
```diff
@@ -9,6 +9,195 @@
 	var/heal_brute = 0
 	var/heal_burn = 0
 
+/obj/item/stack/medical/proc/radial_medical(mob/living/carbon/M as mob, mob/user as mob)
```
Comment (Rohesie) [inline code/game/objects/items/stacks/medical.dm]:
```suggestion
/obj/item/stack/medical/proc/radial_medical(mob/living/carbon/patient, mob/doctor)
```
1) Informative vars.
2) `as type` is only useful for verbs and stuff where the person needs to select from lists and the likes.

File: code/game/objects/items/stacks/medical.dm [hunk 2]
```diff
@@ -9,6 +9,195 @@
...
+	var/mob/living/carbon/human/H = M
```
Comment (Rohesie) [inline code/game/objects/items/stacks/medical.dm]: This can be defined in the proc call, instead of `M`

**Comment by TheVekter** [comment]
>I bet this could work for surgery steps.

I would love you forever. Can we plan like this might happen in the future and move the proc elsewhere for now? If you don't, I might take a crack at it.

**Comment by Hughgent** [comment]
image
it really does!
And it should also be able to take special "in surgery" radial menu sprites

**Comment by Hughgent** [comment]
image
pretty nice, preferences work.

**Comment by Hughgent** [comment]
image
image

**Comment by Rohesie** [comment]
Good feature, but pending being turned into a component.

**Comment by TheBimmer231** [comment]
👍

File: code/modules/client/preferences_savefile.dm [hunk 3]
```diff
@@ -1,514 +1,521 @@
...
-	return TRUE
-
-
-/datum/preferences/proc/save()
-	return (save_preferences() && save_character())
-
-
-/datum/preferences/proc/load()
-	return (load_preferences() && load_character())
+//This is the lowest supported version, anything below this is completely obsolete and the entire savefile will be wiped.
```
Comment (psykzz) [inline code/modules/client/preferences_savefile.dm]: Im assuming you did a big whitespace change on this file, can you revert?

File: code/modules/surgery/surgery.dm [hunk 4]
```diff
@@ -123,9 +123,9 @@ proc/do_surgery(mob/living/carbon/M, mob/living/user, obj/item/tool)
 		var/fumbling_time = SKILL_TASK_FORMIDABLE - ( SKILL_TASK_AVERAGE * user.mind.cm_skills.surgery ) // 20 secs non-trained, 15 amateur, 10 semi-prof
 		if(!do_after(user, fumbling_time, TRUE, M, BUSY_ICON_UNSKILLED))
 			return
-	var/datum/limb/affected = M.get_limb(user.zone_selected)
+	var/datum/limb/affected = user.client.prefs.toggles_gameplay & RADIAL_MEDICAL ? radial_medical(M, user) : M.get_limb(user.zone_selected)
 	if(!affected)
-		return 0
+		return 1
```
Comment (psykzz) [inline code/modules/surgery/surgery.dm]:
```suggestion
		return TRUE
```

File: code/modules/surgery/surgery.dm [hunk 5]
```diff
@@ -123,9 +123,9 @@ proc/do_surgery(mob/living/carbon/M, mob/living/user, obj/item/tool)
...
 	if(affected.in_surgery_op) //two surgeons can't work on same limb at same time
 		to_chat(user, "<span class='warning'>You can't operate on the patient's [affected.display_name] while it's already being operated on.</span>")
 		return 1
```
Comment (psykzz) [inline code/modules/surgery/surgery.dm]:
```suggestion
		return TRUE
```

I know its not yours but your here.

File: code/game/objects/items/stacks/medical.dm [hunk 6]
```diff
@@ -9,6 +9,70 @@
 	var/heal_brute = 0
 	var/heal_burn = 0
 
+proc/radial_medical(mob/living/carbon/patient, mob/doctor)
```
Comment (psykzz) [inline code/game/objects/items/stacks/medical.dm]: Can you make sure this proc is using the full path?

**Comment by Rohesie** [comment]
This one needs to be a component or something more generic. The current implementation is a bit too hacky.

**Comment by spookydonut** [comment]
> 
> 
> This one needs to be a component or something more generic. The current implementation is a bit too hacky.

it doesnt

File: code/modules/client/preferences_savefile.dm [hunk 7]
```diff
@@ -1,514 +1,521 @@
...
+
+//should these procs get fairly long
+//just increase SAVEFILE_VERSION_MIN so it's not as far behind
+//SAVEFILE_VERSION_MAX and then delete any obsolete if clauses
+//from these procs.
+//This only really meant to avoid annoying frequent players
+//if your savefile is 3 months out of date, then 'tough shit'.
+
+/datum/preferences/proc/update_preferences(current_version, savefile/S)
+	if(current_version < 40)
```
Comment (psykzz) [inline code/modules/client/preferences_savefile.dm]:
```suggestion
	if(current_version < 39)
```

Dont change this

**Comment by Hughgent** [comment]
@Rohesie 
As nice as this being a component would be. It doesn't accomplish it's purpose as a component.

As it turns out, a 'feature' of component is that it does not care what else is happening. "when you walk over a tile with squeak, by god it will squeak no matter what." @ninjanomnom (paraphrased).

Thus, when you click a person with a bandage, the radial menu does open up, BUT the rest of the code continues to process. which means you immediately apply the bandage/whatever to your selected attack zone.

The purpose of this as a `proc` is that it forces the code to stop and wait for the persons input when that input is important. (any of the procs where `affecting` is a var.)

From a game play perspective, I want to apply a bandage to a marines head, but I want to keep shooting at the chest area. The component version changes my target zone, but the proc version just applies the bandage to the head without changing the target zone.

**Comment by spookydonut** [comment]
This doesnt need to be a component and @Rohesie is being unreasonable.

**Comment by spookydonut** [comment]
Pending any bugs found I intend to merge this in the next day or so, TM or not.

**Comment by Rohesie** [comment]
This will alter preferences and I'd appreciate if you didn't go over my request. This is the second PR merged doing that.
Please revert both.

**Comment by spookydonut** [comment]
@Rohesie care to point out where you raised these objections?

And what's the two PRs I've merged 'going over your request' because this PR had to demanding an unreasonable change request and slapping DNM labels on the PR with no reasoning, which is something you've repeated accused me of.

**Comment by Rohesie** [comment]
Check this PR and you'll see me adding the Do Not Merge label twice, both removed by you, the last time simply declaring me unreasonable.

We've discussed this extensively on Discord and you've memed me about it yesterday when I was talking about a completely unrelated thing:
https://i.imgur.com/AMJnh1V.png

This is the other PR is this one, same pattern: https://github.com/tgstation/TerraGov-Marine-Corps/pull/3329
We have the discussion on Discord here: https://discordapp.com/channels/498569646014857217/498576390753615882/644736802489499649
And there's the change requests still on that PR.

You seem to be implying you didn't see my objections, but you made clear allusions to them here and elsewhere. Was that what you were trying to say or am I getting the wrong message? And if not, then why are you asking for me to point out where I raised them?

**Comment by spookydonut** [comment]
I can't see what you're linking to on discord because you picked a fight with me on that discord then banned me when I told you to fuck off like I do every time you pick a fight.

Boiler PR was fine, as was this one, you're demanding people make things components without really understanding why except cargo culting, and then not offering any help to the people.

**Comment by spookydonut** [comment]
> Check this PR and you'll see me adding the Do Not Merge label twice, both removed by you, the last time simply declaring me unreasonable.
Both times with no reasoning given

**Pull request merged**